index.d.ts

```diff
@@ -0,0 +1,42 @@
+import { FastifyPluginAsync } from 'fastify'
+import { DirectiveNode, GraphQLResolveInfo } from 'graphql'
+import { MercuriusContext } from 'mercurius'
+
+/**
+ * the policy promise to run when an auth directive protected field is selected by the query.
+ * This must return true in order to pass the check and allow access to the protected field.
+ */
+export type ApplyPolicyHandler<TParent=any, TArgs=any, TContext=MercuriusContext> = (
+  authDirectiveAST: DirectiveNode,
+  parent: TParent,
+  args: TArgs,
+  context: TContext,
+  info: GraphQLResolveInfo
+) => Promise<boolean | Error>;
+
+/** assigns the returned data to `MercuriusContext.auth` for use in the `applyPolicy` function. */
+export type AuthContextHandler<TContext=MercuriusContext> = (context: TContext) => object | Promise<object>;
+
+export interface MercuriusAuthOptions<TParent=any, TArgs=any, TContext=MercuriusContext> {
+  /** The name of the directive that the Mercurius auth plugin will look for within the GraphQL schema in order to identify protected fields. For example, for directive definition `directive @auth on OBJECT | FIELD_DEFINITION`, the corresponding name would be auth. */
+  authDirective: string;
+  /**
+   * the policy promise to run when an auth directive protected field is selected by the query.
+   * This must return true in order to pass the check and allow access to the protected field.
+   */
+  applyPolicy: ApplyPolicyHandler<TParent, TArgs, TContext>
+  /** assigns the returned data to `MercuriusContext.auth` for use in the `applyPolicy` function. */
+  authContext?: AuthContextHandler<TContext>;
+}
+
+/** Mercurius Auth is a plugin for `mercurius` that adds configurable Authentication and Authorization support. */
+declare const mercuriusAuth: FastifyPluginAsync<MercuriusAuthOptions>
+
+declare module 'mercurius' {
+  interface MercuriusAuthContext {
+    auth?: Record<string, any>
+  }
+  export interface MercuriusContext extends MercuriusAuthContext {}
```

Comment:
I think the `MercuriusAuthContext` is more conceptually the type of `auth`. Could we export the MercuriusAuthContext type and avoid the extension of MercuriusContext?

Something like:

```ts
export interface MercuriusAuthContext extends Record<string, any> {}
export interface MercuriusContext {
  auth?: MercuriusAuthContext;
}
```

within the `mercurius` module declaration :)

Reply:
but where will the `MercuriusAuthContext` be used?

Reply:
The return type of the authContext handler and users may want to use to it - I know I will

Reply:
> The return type of the authContext handler and users may want to use to it - I know I will

ok that's done

Reply:
https://github.com/mercurius-js/auth/pull/7#discussion_r630788972